bft/abci/abci.go

```diff
@@ -41,12 +54,166 @@ func (app *VerificationApp) PrepareProposal(_ context.Context, proposal *abcityp
+	for _, slice := range proposal.Txs {
+		var transaction types.Transaction
+		err := proto.Unmarshal(slice, &transaction)
+		if err != nil {
+			log.Error("Error unmarshaling transaction data:", err)
+		}
+
+		switch transaction.Type {
+		case types.TransactionType_VerificationTransaction:
+			result = append(result, slice)
+
+		case types.TransactionType_SummaryTransaction:
+			log.Debug("We are removing this TX")
+		default:
+			othertx = append(othertx, slice)
+		}
+
+	}
+
+	var xoredTx = help.XorArrays(result)
+
+	log.Debug("Merging Done")
+
+	hash := sha256.Sum256(xoredTx)
+	hashString := base64.StdEncoding.EncodeToString(hash[:])
+	transactionMessage := types.Transaction{
+		Owner:     "trial",
+		Signature: "",
+		Type:      *types.TransactionType_SummaryTransaction.Enum(),
+	}
+	transactionMessage.Data = &types.Transaction_SummaryTransactionData{
+		SummaryTransactionData: &types.SummaryTransactionData{
+			Hash:  hashString,
+			NumTx: int64(comettype.ToTxs(proposal.Txs).Len()),
+		},
+	}
+	transactionBytes, err := proto.Marshal(&transactionMessage)
+	if err != nil {
+		panic(err)
+	}
+	log.Debug("Marshaling Done")
+	transactions := comettype.Tx(transactionBytes[:])
+
+	transactionMessage_checkpoint := types.Transaction{
+		Owner:     "trial",
+		Signature: "",
+		Type:      *types.TransactionType_PolygonCheckpointTransaction.Enum(),
+	}
+	transactionMessage_checkpoint.Data = &types.Transaction_PolygonCheckpointTransactionData{
+		PolygonCheckpointTransactionData: &types.PolygonCheckpointTransactionData{
+			Blockno:   uint64(app.Tracker.LastTrackerHeight),
+			Blockhash: "xyz", //not a necessary field just nice to have for record keeping
+		},
+	}
+	transactionBytes_Checkpoint, err := proto.Marshal(&transactionMessage_checkpoint)
+	if err != nil {
+		panic(err)
+	}
+
+	transactions_checkpoint := comettype.Tx(transactionBytes_Checkpoint[:])
+	proposal.Txs = append(othertx, transactions, transactions_checkpoint)
+	log.Debug(proposal.Txs)
 	return &abcitypes.ResponsePrepareProposal{Txs: proposal.Txs}, nil
 }
 func (app *VerificationApp) ProcessProposal(_ context.Context, proposal *abcitypes.RequestProcessProposal) (*abcitypes.ResponseProcessProposal, error) {
 
 	// Supposedly it's bad for performance to reject crappy blocks.
 	// I think we should be a strict as possible, and give death penalty to misbehaving nodes basically.
+	total_tx := app.Node.Mempool().ReapMaxTxs(-1)
+	log.Debug("The size for the node is")
+	log.Debug(app.Node.Mempool().Size())
+	app.CurrentMempool = total_tx
+	if len(app.CurrentMempool) > 2 {
+		sort.Slice(app.CurrentMempool, func(i, j int) bool {
+			return string(app.CurrentMempool[i]) < string(app.CurrentMempool[j])
+		})
+	}
+	log.Debug("Sorting Done")
+	var result [][]byte
+	var othertx = [][]byte{}
+	for _, slice := range app.CurrentMempool {
+		var transaction types.Transaction
+		err := proto.Unmarshal(slice, &transaction)
+		if err != nil {
+			log.Error("Error unmarshaling transaction data:", err)
```

Comment:
Can you really continue with the proposal if there is an error here? Shouldn't you return an error?